### PR TITLE
Require building with GHC 9.10

### PR DESCRIPTION
**Pull request by RyanGlScott**
This is a collection of patches needed to make `singletons`, `singletons-th`, and `singletons-base` build with GHC 9.10 (and in the case of `singletons-{th,base}`, _require_ them to build with GHC 9.10). The two most interesting commits (whose commit messages I have included below) are the ones involving changes to the arities of type families. GHC 9.10 no longer performs arity inference (see the commit messages below), so we need to explicitly use `TypeAbstractions` in more places to give promoted type families the correct arities.

This makes more progress towards https://github.com/goldfirere/singletons/issues/569.

## `singletons`: Support building with GHC 9.10

Because GHC 9.10 no longer performs arity inference in type-level declarations (see
https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0425-decl-invis-binders.rst), we now must use `TypeAbstractions` in certain `singletons` definitions to make GHC 9.10 accept them.

One part of a fix for https://github.com/goldfirere/singletons/issues/566.

## `singletons-th`: Adapt to GHC 9.10's lack of arity inference

GHC 9.10 no longer performs arity inference in type-level declarations (see https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0425-decl-invis-binders.rst), and as it turns out, many of the promoted type families that `singletons-th` generates would have the wrong arity. For instance, `singletons-th` would promote this definition:

```hs
f :: Either a Bool
f = Right True
```

To this type family:

```hs
type F :: Either a Bool
type family F where
  F = Right True
```

With GHC 9.10 or later, however, GHC would conclude that `F` has arity 0, which means that it should not bind any arguments (visible or invisible). The type family equation for `F`, however, only works if `F` has arity 1! This is because the type family equation needs to bind an invisible `@a` argument:

```hs
  F @A = Right @A @Bool True
```

To ensure that type families like `F` have the expected arity, `singletons-th` now uses `TypeAbstractions` in more places to ensure that type family headers bind an appropriate number of type variables, which makes the type families' arities explicit. For instance, `singletons-th` now generates the following code for `F`:

```hs
type F :: Either a Bool
type family F @A where -- Note the @A here, which gives it arity 1
  F = Right True
```

For more details on how this is implemented, see the new `Note [Generating type families with the correct arity]` in `Data.Singletons.TH.Promote`.

A consequence of this change is that the average piece of `singletons-th`–generated code is much more likely to require `TypeAbstractions` than it did before. This explains why we now enable `TypeAbstractions` in almost every module in `singletons-base`.

Fixes https://github.com/goldfirere/singletons/issues/566.

### CODE REVIEW

File: singletons-base/tests/compile-and-dump/Singletons/PatternMatching.golden [hunk 1]
```diff
@@ -594,7 +595,24 @@ Singletons/PatternMatching.hs:(0,0)-(0,0): Splicing declarations
...
+                                          GHC.Types.LiftedRep
+                                          ((:)
+                                             @GHC.Types.RuntimeRep
+                                             GHC.Types.LiftedRep
+                                             ('[] @GHC.Types.RuntimeRep))))’
+    • In a standalone kind signature for ‘Foo2’: (#,#) a b -> a
+   |
+19 | $(singletons [d|
+   |  ^^^^^^^^^^^^^^^...
+
```
Comment (RyanGlScott) [inline singletons-base/tests/compile-and-dump/Singletons/PatternMatching.golden:618]: This error message is unexpected. Upon further inspection, it's GHC fault, as `unboxedTupleTypeName 2` is returning the wrong `Name`, causing `singletons-th` to handle `(#,#)` in an unexpected way. See https://github.com/goldfirere/th-desugar/issues/213, as well as the upstream issue [GHC#24750](https://gitlab.haskell.org/ghc/ghc/-/issues/24750). Happily, that upstream GHC issue has already been fixed in `master`, so once that is backported to GHC 9.10, then this test case should work once more.

File: singletons-th/src/Data/Singletons/TH/Promote.hs [hunk 2]
```diff
@@ -957,6 +966,56 @@ information whatsoever. Until the aformentioned Haddock issue is resolved, we
...
+
+Note [Generating type families with the correct arity]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+As of GHC 9.10, GHC no longer performs arity inference when kind-checking type
+family declarations with standalone kind signatures. This is an important
+consideration when promoting functions with top-level type signatures. For
+example, we would not want to take this definition:
+
+  f :: Maybe a
+  f = Nothing
```
Comment (vagarenko) [inline singletons-th/src/Data/Singletons/TH/Promote.hs:978]: Did you mean to write 
```
f :: Either a Bool
f = Right True
```
here, like in the PR description?

Comment (RyanGlScott) [reply singletons-th/src/Data/Singletons/TH/Promote.hs]: Oops! I wrote this up, changed my mind about what example to use, and then forget to modify this part. Thanks for noticing this—I've changed it to be consistent with the rest of the documentation.

File: singletons-base/tests/compile-and-dump/Singletons/PatternMatching.golden (hunk 1 above)
Comment (RyanGlScott) [reply singletons-base/tests/compile-and-dump/Singletons/PatternMatching.golden]: Happily, GHC 9.10.1 includes a fix for GHC#24750, so this is no longer an issue.

**Pull request merged**